Homework3/MyThreadPool/MyThreadPool.cs

```diff
@@ -0,0 +1,177 @@
+    }
+
+    public MyThreadPool(int numberOfThreads)
+    {
+        if (numberOfThreads <= 0)
+        {
+            throw new ArgumentException("The number of threads cannot be less than or equal to 0");
+        }
+
+        threads = new MyThread[numberOfThreads];
+        taskQueue = new BlockingCollection<Action>();
+        cancellationTokenSource = new();
+
+        for (int i = 0; i < numberOfThreads; ++i)
+        {
+            threads[i] = new MyThread(taskQueue, cancellationTokenSource.Token);
+        }
+    }
+
+    public IMyTask<TResult> Submit<TResult>(Func<TResult> task) 
+    {
+        if (task == null)
+        {
+            throw new ArgumentNullException("Task can't be null");
+        }
+        cancellationTokenSource.Token.ThrowIfCancellationRequested();
+
+        var newTask = new MyTask<TResult>(task, cancellationTokenSource.Token, taskQueue);
+        taskQueue.Add(newTask.Start);
+        return newTask;
+    }
+
+    public void Shutdown()
+    {
+        cancellationTokenSource.Cancel();
+        foreach (var thread in threads)
+        {
+            thread.Join();
+        }
+    }
+
+    public class MyThread
+    {
+        private readonly Thread thread;
+
+        private readonly BlockingCollection<Action> taskQueue;
+
+        private readonly CancellationToken token;
+
+        private bool IsWorking { get; set; }
+
+        public MyThread(BlockingCollection<Action> taskQueue, CancellationToken token)
+        {
+            this.token = token;
+            this.taskQueue = taskQueue;
+            thread = new Thread(Start);
+            thread.Start();
+        }
+
+        private void Start()
+        {
+            while (true)
+            {
+                if (taskQueue.Count > 0)
+                {
+                    if (taskQueue.TryTake(out var task)) 
+                    {
+                        IsWorking = true;
+                        task();
+                        IsWorking = false;
+                    }
+                } 
+                else if (token.IsCancellationRequested)
+                {
+                    break;
+                }
+            }
+        }
+
+        protected internal void Join()
+        {
+            thread.Join();
+        }
+    }
+
+    private class MyTask<TResult> : IMyTask<TResult>
+    {
+        public bool IsCompleted { get; private set; }
+
+        public Func<TResult> Func { get; private set; }
+
+        private Exception? exception = null;
+
+        private TResult result;
+
+        private readonly ManualResetEvent resetEvent = new(false);
+
+        private readonly CancellationToken token;
+
+        private readonly BlockingCollection<Action> nextTasks;
+
+        private readonly BlockingCollection<Action> taskQueue;
+
+        public TResult Result
+        {
+            get
+            {
+                if (!IsCompleted)
+                {
+                    resetEvent.WaitOne();
+                }
```

Comment:
```suggestion
                resetEvent.WaitOne();
```
И всё работает также, но кода меньше :)